server/api/cart.js

```diff
@@ -0,0 +1,153 @@
+        where: {
+          userId: req.user.id
+        }
+      })
+      //get or create the single item being posted (default qty 0)
+      const [cartItem, created] = await CartItem.findOrCreate({
+        where: {
+          artworkId: artworkId,
+          cartId: cart.id
+        }
+      })
+      //update the quantity of the item to match the qty posted
+      //this is done so that the user can add more of an existing item to the cart
+      //and there won't be multiple entries for the same artwork
+      await cartItem.update({ quantity: cartItem.quantity + quantity})
+      res.status(201).send(cartItem)
+    } else {
+      res.sendStatus(401);
+    }
+  } catch (e) {
+    next(e)
+  }
+})
+
+//this route is for removing a single item in the cart at a time
+router.delete('/item/:id', async (req, res, next) => {
+  try {
+    //only accept if authenticated user
+    if (req.user) {
+      //find the user's cart; there should only be one if all the other stuff
+      //we've done is right. If not, we'll see errors...
+      const cart = await Cart.findOne( {
+        where: {
+          userId: req.user.id
+        }
+      })
+      //kill the item if it's in that user's cart
+      const numDeleted = await CartItem.destroy({
+        where: {
+          id: req.params.id,
+          cartId: cart.id
+        }
+      })
+      //send confirmation of delete if one row was deleted.
+      if(numDeleted === 1) res.sendStatus(204)
+      else throw new Error('Something went wrong')
+    } else {
+      res.sendStatus(401);
+    }
+  } catch (e) {
+    console.log(e)
+    next(e)
+  }
+})
+
+//this route is for updating a single item in the cart
+router.put('/item', async (req, res, next) => {
+  try {
+    //only accept if authenticated user
+    if (req.user) {
+      const { artworkId, quantity } = req.body
+      //find the user's cart; there should only be one if all the other stuff
+      //we've done is right. If not, we'll see errors...
+      const cart = await Cart.findOne( {
+        where: {
+          userId: req.user.id
+        }
+      })
+
+      //according to docs, this should return # of updated rows and an array of
+      //the updated rows, but for some reason it returns [undefined, single cart item]
+      //the query runs and updates the right art, but why it doesn't do what it says
+      //is a mystery to me. One theory: docs say model.Update only returns this
+      //format of data - array with # of rows and an array of rows  - with postgres, so
+      //maybe this version of PG is unsupported? either way, we only want to update one
+      //thing, so this works for now.
+      const [useless, cartItem] = await CartItem.update({
+        quantity: quantity
+      }, {
+        where: {
+          artworkId: artworkId,
+          cartId: cart.id
+        },
+        returning: true, // needed for cartItems to be populated
+        plain: true // makes sure that the returned instances are just plain objects
```

Comment:
i've found these updates to be trickier than expected. another approach i've used to get around this is finding the item first then updating, more DB calls but simpler code.

Reply:
Part of the reason I did it this way is that it needs to make sure the cart belongs to the user making the request. For this route, at least, I think you could go directly to the cartItem using the ID and probably include the cart and include the user, but that seems equally tricky to me. Plus, it wouldn't work for adding cartItems, since there are ones that come from localStorage that do not yet have an ID (although it could be argued that the current way cartItems are added is kinda hacky), so I saw some value in consistency between routes.
Probably it was a mistake to make carts at all, since a user really only should have one cart. If we change that and change the hacky way cartItems are added, it could simplify all these routes. Feels like a lot of backtracking, though. Maybe we can discuss in class tonight.